src/common/state/redis.c

```diff
@@ -967,53 +923,75 @@ void redis_task_table_subscribe_callback(redisAsyncContext *c,
-   * subscription. */
-  if (payload->str == NULL) {
-    if (callback_data->done_callback) {
+  if (strcmp(message_type->str, "message") == 0 ||
+      strcmp(message_type->str, "pmessage") == 0) {
+    /* Handle a task table event. Parse the payload and call the callback. */
+    task_table_subscribe_data *data = callback_data->data;
+    /* Read out the information from the payload. */
+    task_id task_id;
+    int state;
+    db_client_id local_scheduler_id;
+    task_spec *spec;
+    parse_task_table_subscribe_callback(payload->str, payload->len, &task_id,
+                                        &state, &local_scheduler_id, &spec);
+    task *task = alloc_task(spec, state, local_scheduler_id);
+    free(spec);
+    /* Call the subscribe callback if there is one. */
+    if (data->subscribe_callback != NULL) {
+      data->subscribe_callback(task, data->subscribe_context);
+    }
+    free_task(task);
+  } else if (strcmp(message_type->str, "subscribe") == 0 ||
+             strcmp(message_type->str, "psubscribe") == 0) {
+    /* If this condition is true, we got the initial message that acknowledged
+     * the subscription. */
+    if (callback_data->done_callback != NULL) {
       task_table_done_callback done_callback = callback_data->done_callback;
       done_callback(callback_data->id, callback_data->user_context);
     }
     /* Note that we do not destroy the callback data yet because the
      * subscription callback needs this data. */
     event_loop_remove_timer(db->loop, callback_data->timer_id);
-    return;
-  }
-  /* Otherwise, parse the task and call the callback. */
-  task_table_subscribe_data *data = callback_data->data;
-
-  task *task = malloc(payload->len);
-  memcpy(task, payload->str, payload->len);
-  if (data->subscribe_callback) {
-    data->subscribe_callback(task, data->subscribe_context);
+  } else {
+    LOG_FATAL(
+        "Unexpected reply type from task table subscribe. Message type is %s.",
+        message_type->str);
   }
-  free_task(task);
 }
 
 void redis_task_table_subscribe(table_callback_data *callback_data) {
   db_handle *db = callback_data->db_handle;
   task_table_subscribe_data *data = callback_data->data;
-  int status = REDIS_OK;
-  if (IS_NIL_ID(data->node)) {
+  /* TASK_CHANNEL_PREFIX is defined in ray_redis_module.c and must be kept in
+   * sync with that file. */
+  char *TASK_CHANNEL_PREFIX = "TT:";
+  int status;
+  if (IS_NIL_ID(data->local_scheduler_id)) {
     /* TODO(swang): Implement the state_filter by translating the bitmask into
      * a Redis key-matching pattern. */
-    status =
-        redisAsyncCommand(db->sub_context, redis_task_table_subscribe_callback,
-                          (void *) callback_data->timer_id,
-                          "PSUBSCRIBE task:*:%d", data->state_filter);
+    status = redisAsyncCommand(
+        db->sub_context, redis_task_table_subscribe_callback,
+        (void *) callback_data->timer_id, "PSUBSCRIBE %s*:%2d",
```

Comment:
I had a skeleton definition for a Redis module for task table subscription, but I think we can remove that now?